RavenDB-14905 Allow the use of unix ticks for time-series bulk-insert

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-14905 Allow the use of unix ticks for time-series bulk-insert

ayende reviewed (commented):
Where is this used?

karmeli87 commented:
This will be used in the Java client, and in other clients as well.
Also needed for the tsbs benchmark, which is written in Go

ppekrol commented:
@karmeli87 can we write a simple test for this?

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesOperation.cs:
```diff
@@ -64,11 +64,27 @@ internal static TimeSeriesOperation Parse(BlittableJsonReaderObject input)
+        private const string TimeSeriesFormat = "TimeFormat";
+        private const string UnixTimeFormatMs = "UnixTimeFormatMs";
+        private const string UnixTimeFormatNs = "UnixTimeFormatNs";
+        private static readonly long _epochTicks = new DateTime(1970, 1, 1).Ticks;
```
Just to be sure, this shouldn't be `new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc).Ticks` ?

karmeli87 replied:
Let's be on the safe side, done.

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesOperation.cs:
```diff
@@ -64,11 +64,27 @@ internal static TimeSeriesOperation Parse(BlittableJsonReaderObject input)
             if (input.TryGet(nameof(Name), out string name) == false || name == null)
                 ThrowMissingProperty(nameof(Name));
 
+            input.TryGet(TimeSeriesFormat, out string format);
```
Hmm... I would prefer to convert this value to enum and if it exists parse it. If it cannot be parsed then we should throw.

karmeli87 replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesOperation.cs:
```diff
@@ -64,11 +64,27 @@ internal static TimeSeriesOperation Parse(BlittableJsonReaderObject input)
         }
 
+        private const string TimeSeriesFormat = "TimeFormat";
+        private const string UnixTimeFormatMs = "UnixTimeFormatMs";
```
UnixTimeInMs?

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesOperation.cs:
```diff
@@ -64,11 +64,27 @@ internal static TimeSeriesOperation Parse(BlittableJsonReaderObject input)
 
+        private const string TimeSeriesFormat = "TimeFormat";
+        private const string UnixTimeFormatMs = "UnixTimeFormatMs";
+        private const string UnixTimeFormatNs = "UnixTimeFormatNs";
```
UnixTimeInNs?

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesOperation.cs:
```diff
@@ -87,9 +103,21 @@ internal static TimeSeriesOperation ParseForBulkInsert(BlittableJsonReaderObject
                     return null; //never hit
                 }
 
+                long time = (long)bjro[0];
```
This cannot be LazyNumberValue?

karmeli87 replied:
LazyNumberValue has an implicit cast to `long`

ayende replied:
But that won't trigger here. That is something that the _compiler_ inserts, not runtime.

karmeli87 committed:
RavenDB-14905 Create an enum for bulk-instert time format

karmeli87 committed:
RavenDB-14905 address PR comments

- Make sure the time is properly converted to long
- TimeSeriesFormat isn't required

ppekrol committed:
RavenDB-14905 adjustments

karmeli87 commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesOperation.cs:
```diff
@@ -133,6 +171,16 @@ internal static TimeSeriesOperation ParseForBulkInsert(BlittableJsonReaderObject
 
             return result;
+
+            static long GetLong(object value)
```
Will this box? should we `Inline` this?

ayende replied:
No, this shouldn't cause boxing. Why do you expect it to?
